ESP32/main/src/gamepads/ps_3_4_controller.cpp

```diff
@@ -0,0 +1,540 @@
+        stick_ry = stick_ry - sgn(stick_ry) * offset_val;
+        motor.mspeed1 = stick_ry * 5 * global_speed;
+        if (!motor.getEnableMotor())
+          motor.setEnableMotor(true);
+      }
+      else if (motor.mspeed1 != 0) {
+        motor.mspeed1 = 0;
+        motor.stepper_X->setSpeed(motor.mspeed1); // set motor off only once to not affect other modes
+      }
+#endif
+      /*
+          // fine control for Z using buttons
+          if ( PS4.data.button.down) {
+            // fine focus +
+            //run_motor(0, 0, 10);
+            delay(100);
+          }
+          if ( PS4.data.button.up) {
+            // fine focus -
+            //run_motor(0, 0, -10);
+            delay(100);
+          }
+      */
+
+
+#ifdef IS_ANALOG
+      /*
+         Keypad left
+      */
+      if (left()) {
+        // fine lens -
+        analog_val_1 -= 1;
+        delay(100);
+        ledcWrite(analog.PWM_CHANNEL_analog_1, analog_val_1);
+      }
+      if (right()) {
+        // fine lens +
+        analog_val_1 += 1;
+        delay(100);
+        ledcWrite(analog.PWM_CHANNEL_analog_1, analog_val_1);
+      }
+      //unknown button
+      /*if (PS4.data.button.start) {
+        // reset
+        analog_val_1 = 0;
+        ledcWrite(analog->PWM_CHANNEL_analog_1, analog_val_1);
+      }*/
+
+      int offset_val_shoulder = 5;
+      if ( abs(r2()) > offset_val_shoulder) {
+        // analog_val_1++ coarse
+        if ((analog_val_1 + 1000 < pwm_max)) {
+          analog_val_1 += 1000;
+          ledcWrite(analog.PWM_CHANNEL_analog_1, analog_val_1);
+        }
+        if (DEBUG) Serial.println(analog_val_1);
+        delay(100);
+      }
+
+      if ( abs(l2()) > offset_val_shoulder) {
+        // analog_val_1-- coarse
+        if ((analog_val_1 - 1000 > 0)) {
+          analog_val_1 -= 1000;
+          ledcWrite(analog.PWM_CHANNEL_analog_1, analog_val_1);
+        }
+        if (DEBUG) Serial.println(analog_val_1);
+        delay(100);
+      }
+
+
+      if ( abs(r1()) > offset_val_shoulder) {
+        // analog_val_1 + semi coarse
+        if ((analog_val_1 + 100 < pwm_max)) {
+          analog_val_1 += 100;
+          ledcWrite(analog.PWM_CHANNEL_analog_1, analog_val_1);
+          delay(100);
+        }
+      }
+      if ( abs(l1()) > offset_val_shoulder) {
+        // analog_val_1 - semi coarse
+        if ((analog_val_1 - 100 > 0)) {
+          analog_val_1 -= 100;
+          ledcWrite(analog.PWM_CHANNEL_analog_1, analog_val_1);
+          delay(50);
+        }
+      }
+
+#endif
+
+#ifdef IS_MOTOR
+      // run all motors simultaneously
+      motor.stepper_X->setSpeed(motor.mspeed1);
+      motor.stepper_Y->setSpeed(motor.mspeed2);
+      motor.stepper_Z->setSpeed(motor.mspeed3);
+
+      if (motor.mspeed1 || motor.mspeed2 || motor.mspeed3) {
+        motor.isforever = true;
+      }
+      else {
+        motor.isforever = false;
+      }
+#endif
+    }
+
+}
+
+void ps_3_4_controller::ps_onAttach(void * parameter)
```

Comment:
What do the following lines do in the code? I couldn'T find their existence anywhere but here.

Reply:
its an internal wrapper func that can get attached to the psx libs to listen to the psx controller events